### PR TITLE
feat: add support for never expiring PATs

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: Title
Ticket: ME-59

### CODE REVIEW

**Review by kjaskiewiczz — APPROVED** [review]
thanks for taking care of this!
our unit tests don't like the idea of never expiring PATs ;)

**Comment by tranchitella** [comment]
> thanks for taking care of this! our unit tests don't like the idea of never expiring PATs ;)

Yep, I was too optimistic.
I reworked the unit tests and fixed the LICENSE.

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline --fast

**Comment by tranchitella** [comment]
@kjaskiewiczz please review again, I changed the implementation: we now support JWT tokens without the expiration time claim. This is a much nicer (though more intrusive) change and plays nicely with the UI. @mzedel FYI

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline --fast

**Comment by mzedel** [comment]
> we now support JWT tokens without the expiration time claim.

Thanks for doing this! With this in place, do you think going a bit further down [this path](https://github.com/mendersoftware/useradm/pull/339) would allow for the "Stay logged in" checkbox to purely rely on the backend?

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline --fast

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline --fast

File: store/mongo/datastore_mongo.go [hunk 1]
```diff
@@ -431,8 +433,48 @@ func (db *DataStoreMongo) SaveToken(ctx context.Context, token *jwt.Token) error
...
+}
 
-	return nil
+func (db *DataStoreMongo) EnsureSessionTokensLimit(ctx context.Context, userID oid.ObjectID,
+	tokensLimit int) error {
+	opts := &mopts.FindOptions{}
+	opts.SetLimit(int64(tokensLimit))
+	opts.SetSkip(int64(tokensLimit) - 1)
+	opts.SetSort(bson.M{
+		DbTokenIssuedAtTime: -1,
```
Comment (merlin-northern) [inline store/mongo/datastore_mongo.go:445]: ok I checked this one (for off-by one which is always killing me). we have the newest token first, we skip `tokensLimit-1` and then we remove all that are older than the 10th one.

**Review by merlin-northern — COMMENTED** [review]
feel free to ignore, as I was late to the party.

File: store/mongo/datastore_mongo.go [hunk 2]
```diff
@@ -431,8 +433,48 @@ func (db *DataStoreMongo) SaveToken(ctx context.Context, token *jwt.Token) error
...
+	_, err = db.client.
+		Database(mstore.DbFromContext(ctx, DbName)).
+		Collection(DbTokensColl).
+		DeleteMany(ctx, mstore.WithTenantID(ctx, bson.M{
+			DbTokenSubject: userID,
+			DbTokenName:    bson.M{"$exists": false},
+			DbTokenIssuedAtTime: bson.M{
+				"$lt": tokens[0].IssuedAt.Time,
+			},
+		}))
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply store/mongo/datastore_mongo.go:475]: These are session tokens, not tokens that the user "really wants to keep". Those are Personal Access Tokens, you have 10, you name them and use them programmatically. Session tokens are per-session, in the browser (or through APIs) and you are supposed to handle the 401 and log in again if the session is not valid anymore.

Limiting the number of active sessions makes sense independently from the technicalities in the database (where/how we store the JWT tokens).

Regarding the tokens with exactly the same issuedAt, in this case we'll remove them all (or none of them); however, this won't make any significant impact on anything else.

File: api/http/api_useradm.go [hunk 3]
```diff
@@ -156,6 +156,10 @@ func (u *UserAdmApiHandlers) HealthHandler(w rest.ResponseWriter, r *rest.Reques
 	w.WriteHeader(http.StatusNoContent)
 }
 
+type LoginOptions struct {
+	StayLoggedIn bool `json:"stay_logged_in,omitempty"`
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/api_useradm.go]: renamed to `no_expiry`

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline --fast

File: config/config.go [hunk 4]
```diff
@@ -52,6 +52,9 @@ const (
 	SettingDbUsername = "mongo_username"
 	SettingDbPassword = "mongo_password"
 
+	SettingLimitSessionsPerUser        = "limit_sessions_per_user"
+	SettingLimitSessionsPerUserDefault = 10
```
Comment (merlin-northern) [inline config/config.go:56]: do we need an entry here: https://github.com/mendersoftware/useradm/blob/master/config.yaml ?

**Review by merlin-northern — APPROVED** [review]
lgtm, last two comments and sorry for being annoying.

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline --fast

**Pull request merged**